File: longer_but_fast.py
#!/usr/bin/env python3

import sys


def winning_die(enemy_die):
    dim = len(enemy_die)
    total = sum(enemy_die)
    win = [0 for _ in range(total + 1)]
    for i in range(1, total + 1):
        for j in enemy_die:
            win[i] += (i > j) - (j > i)
    subanswers = [[(0, 0) for tot in range(total + 1)] for die in range(dim + 1)]

    for num_die in range(0, dim):
        for subsum in range(num_die + 1, total + 1):
            possibilities = []
            for last in range(1, subsum - num_die + 1):
                poss = (win[last] + subanswers[num_die][subsum - last][0], last)
                possibilities.append(poss)
            subanswers[num_die + 1][subsum] = max(possibilities)
    ans = []
    if subanswers[dim][total][0] > 0:
        cur = total
        for i in range(dim):
            side = subanswers[dim - i][cur][1]
            ans.append(side)
            cur -= side
    return ans


def main():
    numbers = [int(x.strip()) for x in sys.argv[1].split(',')]

    print(','.join([str(n) for n in winning_die(numbers)]))


if __name__ == '__main__':
    main()

File: solutions/java/LongerButFast.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class LongerButFast {

    private static final Pattern PATTERN = Pattern.compile("\\Q,\\E");
    private static final Comparator<int[]> COMPARATOR =
            Comparator.comparing((int[] i) -> Integer.valueOf(i[0])).thenComparing(i -> Integer.valueOf(i[1]));
    private static final int[] EMPTY = new int[2];

    public static void main(final String... args) {
        final int[] theirs = create(args[0]);
        if (theirs == null) {
            throw new IllegalArgumentException();
        }
        final int[] result = LongerButFast.winningDie(theirs);
        if (result == null) {
            System.out.println();
        } else {
            System.out.println(toString(result));
        }
    }

    private static int[] create(final String dice) {
        final String[] sides = PATTERN.split(dice);
        return Arrays.stream(sides)
                .mapToInt(s -> Integer.parseInt(s))
                .toArray();
    }

    public static String toString(final int[] die) {
        return IntStream.of(die)
                .mapToObj(String::valueOf)
                .collect(Collectors.joining(","));
    }

    private static int[] winningDie(final int[] enemy) {
        final int dim = enemy.length;
        final int total = IntStream.of(enemy).sum();
        final int[] win = new int[total + 1];
        for (int i = 1; i <= total; i++) {
            for (final int j : enemy) {
                win[i] += ((i > j) ? 1 : 0) - ((j > i) ? 1 : 0);
            }
        }
        final Subanswers subanswers = new Subanswers();
        for (int dieNum = 0; dieNum < dim; dieNum++) {
            final int curDieNum = dieNum;
            for (int subsum = dieNum + 1; subsum <= total; subsum++) {
                final int curSubSum = subsum;
                final int[] possibility = IntStream.range(1, curSubSum - curDieNum + 1)
                        .mapToObj(last -> new int[]{win[last] + subanswers.get(curDieNum, curSubSum - last)[0], last})
                        .max(COMPARATOR).orElseThrow(IllegalStateException::new);
                subanswers.set(curDieNum + 1, curSubSum, possibility);
            }
        }
        if (subanswers.get(dim, total)[0] > 0) {
            final int[] ans = new int[dim];
            int cur = total;
            for (int i = 0; i < dim; i++) {
                final int side = subanswers.get(dim - i, cur)[1];
                ans[i] = side;
                cur -= side;
            }
            return ans;
        } else {
            return null;
        }
    }

    private static final class Subanswers {

        private final Map<Integer, Map<Integer, int[]>> subanswers = new HashMap<>();

        public void set(final int dieNum, final int subsum, final int[] values) {
            final Integer die = Integer.valueOf(dieNum);
            subanswers.computeIfAbsent(die, num -> new HashMap<>());
            final Map<Integer, int[]> subsums = subanswers.get(die);
            subsums.put(subsum, values);
        }

        public int[] get(final int dieNum, final int subsum) {
            final Integer die = Integer.valueOf(dieNum);
            if (!subanswers.containsKey(die)) {
                return EMPTY;
            }
            final Map<Integer, int[]> subsums = subanswers.get(die);
            return subsums.getOrDefault(subsum, EMPTY);
        }
    }
}
